SourceCode/raytracer/raytracer/KDTreeNode.cpp

```diff
@@ -0,0 +1,135 @@
+#include "KDTreeNode.h"
+
+#include "json.hpp"
+
+#include "TraceHit.h"
+#include "CRTTypes.h"
+#include "Scene.h"
+
+using ordered_json = nlohmann::ordered_json;
+
+void KDTreeNode::traverse(const Scene& scene, const Ray& ray, TraceHit& out) const {
+    out.t = std::numeric_limits<float>::max();
+    out.type = TraceHitType::OUT_OF_BOUNDS;
+
+    traverseRecursive(scene, ray, out);
+}
+
+// TODO optimize triangle intersections by sorting by ray - plane intersection distance first
+void KDTreeNode::traverseRecursive(const Scene& scene, const Ray& ray, TraceHit& out) const {
+    // Early out
+    if (!aabb.hasIntersection(ray)) {
+        return;
+    }
+
+    if (isLeaf()) {
+        // TODO: refactor this into: intersectTriangles(scene, ray, triangleRefs, out);
+        for (const size_t& triRef : triangleRefs) {
+            const Triangle& tri = scene.triangles[triRef];
+
+            TraceHit tryHit{};
+            tri.intersect(scene, ray, triRef, tryHit);
+            if (tryHit.successful() && tryHit.t < out.t && aabb.contains(tryHit.p)) {
+                out = tryHit;
+            }
+        }
+    }
+    else {
+        std::array<const KDTreeNode*, 2> sortedChildren = closestChildren(ray.origin);
+
+        for (size_t i = 0; i < sortedChildren.size(); ++i) {
+            const KDTreeNode* childPtr = sortedChildren[i];
+            childPtr->traverseRecursive(scene, ray, out);
+
+            if (out.successful()) {
+                break;
+            }
+        }
+    }
+}
+
+void KDTreeNode::processAabbDebug(const Scene& scene, size_t childIdx, TraceHit& out) const
+{
+    scene; childIdx, out;
+}
+
+std::array<const KDTreeNode*, 2> KDTreeNode::closestChildren(const Vec3& point) const {
+    assert(axisSplit >= 0 && axisSplit <= 2);
+
+    float dist0 = child[0]->aabb.distanceToAxis(axisSplit, point);
+    float dist1 = child[1]->aabb.distanceToAxis(axisSplit, point);
+    if (dist0 < dist1) {
+        return { child[0].get(), child[1].get() };
+    }
+    else {
+        return { child[1].get(), child[0].get() };
+    }
+}
+
+void KDTreeNode::build(std::vector<size_t>&& newTriangleRefs, const std::vector<AABB>& triangleAABBs, size_t maxTrianglesPerLeaf, size_t maxDepth, size_t depth)
```

Comment:
It would be better to have separate Acceleration Structure class, which manages the build, not the Node directly.